### PR TITLE
Port `sort` to structured kernels.

### PR DESCRIPTION
**Pull request by ysiraichi**
Tracking Issue: #55070

This PR relands #67016, with the modifications discussed in https://github.com/pytorch/pytorch/pull/67015#issuecomment-982004500.

In summary, we call `infer_dense_strides` on the input's strides, and pass it to `set_output`. Meaning that if one of the outputs is resized (by a `resize_output` call), we will also restride such an output using the dense strides of the input.

### CODE REVIEW

File: aten/src/ATen/native/Sorting.cpp [hunk 1]
```diff
@@ -14,30 +14,68 @@
...
+  auto values_defined = values.defined();
+  const auto& indices = maybe_get_output(1);
+  auto indices_defined = indices.defined();
+
+  std::vector<int64_t> strides = (self.is_non_overlapping_and_dense())
+      ? self.strides().vec()
+      : at::infer_dense_strides(self.sizes(), self.strides());
+
+  // 'set_output' will only modify the strides, if the output was resized.
+  // Which means that we can ignore the output's strides here.
```
Comment (bdhirsh) [inline aten/src/ATen/native/Sorting.cpp:56]: is this comment out of date? (strides aren't ignored)

Comment (bdhirsh) [reply aten/src/ATen/native/Sorting.cpp:56]: ohh "ignore the output's strides here" is probably referring to whatever `maybe_get_output()[0].strides()` currently is, when we're in the out= case.

nvm, i think that comment is phrased fine.

File: aten/src/ATen/native/cuda/Sort.cpp [hunk 2]
```diff
@@ -60,37 +63,9 @@ std::tuple<Tensor &,Tensor &> sort_out_stable_cuda(const Tensor & self, c10::opt
   TORCH_CHECK(self_dtype != ScalarType::ComplexFloat && self_dtype != ScalarType::ComplexDouble,
```
Comment (bdhirsh) [inline aten/src/ATen/native/cuda/Sort.cpp:63]: welp, more cuda-only checks. There probably isn't much we can do about these though... so I think it's fine to leave them here.

File: aten/src/ATen/native/cuda/Sort.cpp [hunk 3]
```diff
@@ -161,20 +111,11 @@ std::tuple<Tensor &,Tensor &> sort_out_stable_cuda(const Tensor & self, c10::opt
...
-  return sort_out_stable_cuda(self, stable, dim, descending, values, indices);
-}
-
-std::tuple<Tensor,Tensor> sort_cuda(const Tensor & self, int64_t dim, bool descending) {
-  return sort_stable_cuda(self, /*stable=*/false, dim, descending);
-}
+// TODO: we should handle this accordingly when we start using REGISTER_HIP_DISPATCH,
+// since REGISTER_DISPATCH won't work in this cpp file.
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+REGISTER_CUDA_DISPATCH(sort_stub, &sort_cuda_kernel);
```
Comment (bdhirsh) [inline aten/src/ATen/native/cuda/Sort.cpp:119]: where did this come from?

File: test/test_sort_and_select.py [hunk 4]
```diff
@@ -167,6 +167,28 @@ def test_sort_large(self, device, dtype):
...
+        inputs = [
+            torch.randn(shape, dtype=dtype, device=device),
+            torch.zeros(shape, dtype=dtype, device=device),
+            torch.zeros(shape, dtype=torch.long, device=device),
+        ]
+        # Input: non-contiguous (stride: 5) 3-element array
+        tensor = inputs[0][:, 0]
+        # Outputs: non-contiguous (stride: 5) 2-element array
+        values = inputs[1][0:2, 0]
+        indices = inputs[2][0:2, 0]
```
Comment (bdhirsh) [inline test/test_sort_and_select.py:183]: Hmmm.. will this test the same behavior from the original [bug](https://github.com/pytorch/pytorch/issues/65863)? I think so but I'm not sure.

In the original repro example, the two output tensors were actually too small, vs. here they are views into a larger tensor of the right size. Might be worth just making this test more similar to the repro?

**Comment by bdhirsh** [comment]
Just left some minor comments but looks good! Thanks for working on re-landing this!

File: aten/src/ATen/native/Sorting.cpp (hunk 1 above)
Comment (ysiraichi) [reply aten/src/ATen/native/Sorting.cpp:56]: Nah. This comment wasn't very well worded. I will change it explaining the `infer_dense_strides` call.

File: aten/src/ATen/native/cuda/Sort.cpp [hunk 5]
```diff
@@ -39,15 +40,17 @@ void fillSliceWithIndex(Tensor& t,int dim) {
...
+    const Tensor& self,
+    int64_t dim,
+    bool descending,
+    bool stable,
+    const Tensor& values,
+    const Tensor& indices) {
   // this algorithm is always stable
-  TORCH_INTERNAL_ASSERT(stable.has_value(), "sort_out(): c10::optional<bool> for stable has to have value.");
-  TensorArg self_arg{self, "self", 1}, values_arg{values, "values", 2}, indices_arg{indices, "indices", 3};
-  checkAllSameGPU(__func__, {self_arg, values_arg, indices_arg});
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply aten/src/ATen/native/cuda/Sort.cpp:46]: - `TORCH_INTERNAL_ASSERT(stable.has_value(), ...)`: already being checked on `META`
- `checkAllSameGPU(...)`: already checked via codegen

File: aten/src/ATen/native/cuda/Sort.cpp (hunk 3 above)
Comment (ysiraichi) [reply aten/src/ATen/native/cuda/Sort.cpp:119]: I don't think I got what you mean with this question. I needed this, since I am using `DispatchStub`, here.

File: test/test_sort_and_select.py (hunk 4 above)
Comment (ysiraichi) [reply test/test_sort_and_select.py:183]: You are right. I did it like this because I wanted to make it explicit that the strides weren't `(1,)` before. But, I guess that's not really necessary. Will change it.

File: aten/src/ATen/native/Sorting.cpp [hunk 6]
```diff
@@ -14,30 +14,68 @@
...
   }
+
+  if (indices_defined) {
+    at::assert_no_internal_overlap(indices);
+    at::assert_no_overlap(self, indices);
+  }
+
+  if (values_defined && indices_defined) {
+    at::assert_no_overlap(values, indices);
+  }
```
Comment (ysiraichi) [inline aten/src/ATen/native/Sorting.cpp:76]: Do you think these overlap checks make sense? The `sort` kernel didn't really check for them before, but I thought I would introduce, anyway.

Comment (bdhirsh) [reply aten/src/ATen/native/Sorting.cpp:77]: oh, thanks for calling that out.

That seems like a pretty reasonable check, since otherwise you'd get silently incorrect results.

I would still make it a separate PR though, that way there aren't any unnecessarily "functionality changes" in the port itself. Do you mind making it a separate PR after this one in the stack

File: aten/src/ATen/native/cuda/Sort.cpp (hunk 3 above)
Comment (bdhirsh) [reply aten/src/ATen/native/cuda/Sort.cpp:119]: Just looked at this a little more. Is using `DispatchStub` for the cuda kernel actually necessary? My understanding of `DispatchStub` is:
- it allows you to do dynamic dispatch (technically redundant, since the dispatcher lets you do this too)
- it re-compiles the cpu kernel with multiple vector instruction sets

So the `sort` kernel is already using `DispatchStub` in the cpu kernel (which results in having to do 2 dynamic dispatches - one to go to CPU, and a second to go to the correct CPU vector instruction set).

But it looks like the cuda kernel in master today doesn't bother with dispatch stub. We use the dispatcher to dynamically dispatch to the cuda kernel and that's it. If the `REGISTER_CUDA_DISPATCH` is important for some other functionality (like hip support), then we can probably save that for a future PR, since it will also result in a second dynamic dispatch in the cuda path.

Lmk what you think!

Comment (ysiraichi) [reply aten/src/ATen/native/cuda/Sort.cpp:119]: It's true that I didn't really need `DispatchStub`. That said, I think it's nice here since we get to unify common `CPU` and `CUDA` implementation, in exchange of a little overhead (I think). Do you think it's better to duplicate the code?

> it re-compiles the cpu kernel with multiple vector instruction sets

I thought that every source file inside `native/cpu` got compiled with different instructions sets (not really dispatched to them, though). Am I missing something?

Comment (bdhirsh) [reply aten/src/ATen/native/cuda/Sort.cpp:119]: Ok, I think it's pretty reasonable (you're right, and we make that same tradeoff in a bunch of other ops that are implemented with DispatchStub). Sounds good!

> I thought that every source file inside native/cpu got compiled with different instructions sets (not really dispatched to them, though). Am I missing something?

Yep, that's my understanding too. And then the `sort_stub()` call is technically a dynamic dispatch that will pick the "correct" pre-compiled vectorized cpu kernel (in addition to potentially picking the cuda kernel), based on whatever hardware you happening to be using when running the op.

**Review by bdhirsh — APPROVED** [review]
LGTM! (XLA test failure looks flaky so I kicked it off again).

File: aten/src/ATen/native/cuda/Sort.cpp [hunk 7]
```diff
@@ -39,15 +40,17 @@ void fillSliceWithIndex(Tensor& t,int dim) {
...
 
   bool is_non_overlapping_and_dense = self.is_non_overlapping_and_dense();
   int64_t ndim = self.dim();
```
Comment (bdhirsh) [inline aten/src/ATen/native/cuda/Sort.cpp:53]: internal tests are complaining about the unused variable here

**Comment by ezyang** [comment]
just waitin for it to build :>

**Comment by ysiraichi** [comment]
@ezyang I think the tests are finally passing. There is a failure, though: `Meta Internal-Only Changes Check`. Should I do something about it?

**Comment by janeyx99** [comment]
Adding the label ciflow/trunk should run that failing build in your reland PR!

**Comment by ezyang** [comment]
```
/var/lib/jenkins/workspace/aten/src/ATen/native/cuda/TensorTopK.cpp: In member function ‘void at::native::structured_topk_out_cuda::impl(const at::Tensor&, int64_t, int64_t, bool, bool, const at::Tensor&, const at::Tensor&)’:
/var/lib/jenkins/workspace/aten/src/ATen/native/cuda/TensorTopK.cpp:86:17: error: ‘sort_outf’ is not a member of ‘at::cuda’
       at::cuda::sort_outf(values, /* stable= */ false, dim, largest, sortedValues, sortedIndices);
                 ^~~~~~~~~
```

prolly missing a header include
